In [1]:
import pandas as pd
import numpy as np

In [2]:
from google.colab import files
data=files.upload()

In [3]:
df=pd.read_csv('/content/train.csv')

In [4]:
df.head()

User_ID Product_ID Gender   Age  Occupation City_Category  \
0  1000001  P00069042      F  0-17          10             A   
1  1000001  P00248942      F  0-17          10             A   
2  1000001  P00087842      F  0-17          10             A   
3  1000001  P00085442      F  0-17          10             A   
4  1000002  P00285442      M   55+          16             C   

  Stay_In_Current_City_Years  Marital_Status  Product_Category_1  \
0                          2             0.0                 3.0   
1                          2             0.0                 1.0   
2                          2             0.0                12.0   
3                          2             0.0                12.0   
4                         4+             0.0                 8.0   

   Product_Category_2  Product_Category_3  Purchase  
0                 NaN                 NaN    8370.0  
1                 6.0                14.0   15200.0  
2                 NaN                 NaN    1422.0  
3                14.0                 NaN    1057.0  
4                 NaN                 NaN    7969.0

In [5]:
df.shape

(203282, 12)

#1.	How do we convert "4+" in "Stay_in_current_city_years" variable  to a value of 4?

In [6]:
df['Stay_In_Current_City_Years'] = df['Stay_In_Current_City_Years'].replace(['4+'], '4')

In [7]:
df.head(20)

User_ID Product_ID Gender    Age  Occupation City_Category  \
0   1000001  P00069042      F   0-17          10             A   
1   1000001  P00248942      F   0-17          10             A   
2   1000001  P00087842      F   0-17          10             A   
3   1000001  P00085442      F   0-17          10             A   
4   1000002  P00285442      M    55+          16             C   
5   1000003  P00193542      M  26-35          15             A   
6   1000004  P00184942      M  46-50           7             B   
7   1000004  P00346142      M  46-50           7             B   
8   1000004   P0097242      M  46-50           7             B   
9   1000005  P00274942      M  26-35          20             A   
10  1000005  P00251242      M  26-35          20             A   
11  1000005  P00014542      M  26-35          20             A   
12  1000005  P00031342      M  26-35          20             A   
13  1000005  P00145042      M  26-35          20             A   
14  1000006  P00231342      F  51-55           9             A   
15  1000006  P00190242      F  51-55           9             A   
16  1000006   P0096642      F  51-55           9             A   
17  1000006  P00058442      F  51-55           9             A   
18  1000007  P00036842      M  36-45           1             B   
19  1000008  P00249542      M  26-35          12             C   

   Stay_In_Current_City_Years  Marital_Status  Product_Category_1  \
0                           2             0.0                 3.0   
1                           2             0.0                 1.0   
2                           2             0.0                12.0   
3                           2             0.0                12.0   
4                           4             0.0                 8.0   
5                           3             0.0                 1.0   
6                           2             1.0                 1.0   
7                           2             1.0                 1.0   
8                           2             1.0                 1.0   
9                           1             1.0                 8.0   
10                          1             1.0                 5.0   
11                          1             1.0                 8.0   
12                          1             1.0                 8.0   
13                          1             1.0                 1.0   
14                          1             0.0                 5.0   
15                          1             0.0                 4.0   
16                          1             0.0                 2.0   
17                          1             0.0                 5.0   
18                          1             1.0                 1.0   
19                          4             1.0                 1.0   

    Product_Category_2  Product_Category_3  Purchase  
0                  NaN                 NaN    8370.0  
1                  6.0                14.0   15200.0  
2                  NaN                 NaN    1422.0  
3                 14.0                 NaN    1057.0  
4                  NaN                 NaN    7969.0  
5                  2.0                 NaN   15227.0  
6                  8.0                17.0   19215.0  
7                 15.0                 NaN   15854.0  
8                 16.0                 NaN   15686.0  
9                  NaN                 NaN    7871.0  
10                11.0                 NaN    5254.0  
11                 NaN                 NaN    3957.0  
12                 NaN                 NaN    6073.0  
13                 2.0                 5.0   15665.0  
14                 8.0                14.0    5378.0  
15                 5.0                 NaN    2079.0  
16                 3.0                 4.0   13055.0  
17                14.0                 NaN    8851.0  
18                14.0                16.0   11788.0  
19                 5.0                15.0   19614.0

#2.	How many rows exist with a marital_status = 0?

In [10]:
print(len(df))
print(len(df.columns))

203282
12


In [11]:
(df['Marital_Status']== 0).sum()

119939



so totally 119939 rows having the '0's

#3.	How many rows exist within the age_group of "26-35" with a marital status =0?

In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 203282 entries, 0 to 203281
Data columns (total 12 columns):
 #   Column                      Non-Null Count   Dtype  
---  ------                      --------------   -----  
 0   User_ID                     203282 non-null  int64  
 1   Product_ID                  203282 non-null  object 
 2   Gender                      203282 non-null  object 
 3   Age                         203282 non-null  object 
 4   Occupation                  203282 non-null  int64  
 5   City_Category               203282 non-null  object 
 6   Stay_In_Current_City_Years  203282 non-null  object 
 7   Marital_Status              203281 non-null  float64
 8   Product_Category_1          203281 non-null  float64
 9   Product_Category_2          140237 non-null  float64
 10  Product_Category_3          62257 non-null   float64
 11  Purchase                    203281 non-null  float64
dtypes: float64(5), int64(2), object(5)
memory usage: 18.6+ MB


In [15]:
a = ((df['Marital_Status']== 0) & (df['Age']== 26- 35)) .sum()
a

0

#4.	How many distinct users exist within the age_group of "26-35" with a marital status =0?

In [16]:
distinct_users = df[(df['Marital_Status']== 0) & (df['Age']== 26-35)]
nousers =distinct_users['User_ID'].nunique()
nousers

0

so there is NO distinct user exits in the age group of "26-35" with a marital status =0

#5.How many distinct age groups exist?

In [18]:
age_group=df['Age'].unique()
age_group

array(['0-17', '55+', '26-35', '46-50', '51-55', '36-45', '18-25'],
      dtype=object)

so there is 7 different age group exits in the train dataset.

#6.How many distinct User_IDs exist?

In [19]:
user_id = df['User_ID'].nunique()
user_id

5888

so,totally 5888 unique user-id exits in train dataset.

#7.Which product_ID occurs the most frequently?

In [20]:
pro_id = df['Product_ID'].value_counts()
pro_id

P00025442    634
P00265242    628
P00112142    608
P00110742    606
P00057642    564
            ... 
P00167242      1
P00069942      1
P00228842      1
P00143242      1
P00169242      1
Name: Product_ID, Length: 3469, dtype: int64

this is the product-id is most frquently occured in the train dataset.

#8.	What is the average purchase rate of gender = f & gender = m?

In [22]:
avg = df.groupby('Gender')['Purchase'].mean()
avg

Gender
F    8808.991046
M    9485.173848
Name: Purchase, dtype: float64

Gender
F    8808.991046
M    9485.173848 over all purchase of female and male.

#9.	What is the average value of purchase when gender = f or age_group = "0-17"

In [23]:
average_purchase = df[(df["Gender"] == "F") | (df["Age"] == "0-17")]["Purchase"].mean()
print(average_purchase)

8846.101254996745


#10.	What is the average value of purchase within the odd rows of train.csv?

In [24]:
avg_odd = df[df.index % 2 != 0]["Purchase"].mean()
avg_odd

9329.529476584023

9329.529476584023 so totally average value of purchase within the odd rows of train dataset

#11.	create a new dataset (train2) that does not have any row in train.csv that has  missing value

In [25]:
train2 = df.dropna()
train2.head()

User_ID Product_ID Gender    Age  Occupation City_Category  \
1   1000001  P00248942      F   0-17          10             A   
6   1000004  P00184942      M  46-50           7             B   
13  1000005  P00145042      M  26-35          20             A   
14  1000006  P00231342      F  51-55           9             A   
16  1000006   P0096642      F  51-55           9             A   

   Stay_In_Current_City_Years  Marital_Status  Product_Category_1  \
1                           2             0.0                 1.0   
6                           2             1.0                 1.0   
13                          1             1.0                 1.0   
14                          1             0.0                 5.0   
16                          1             0.0                 2.0   

    Product_Category_2  Product_Category_3  Purchase  
1                  6.0                14.0   15200.0  
6                  8.0                17.0   19215.0  
13                 2.0                 5.0   15665.0  
14                 8.0                14.0    5378.0  
16                 3.0                 4.0   13055.0

#12.	In which city_category do most of the users within age group "0-17" live?

In [26]:
city =df[(df["Age"] == "0-17")]["City_Category"].value_counts().index[0]
city

'C'

#13.	For how many rows is "Product_category_2" missing a value?

In [27]:
missing_product = df['Product_Category_2'].isna().sum()
missing_product

63045

so, totally 63045 values are missed in the product category2

#14.	Which value of "Product_category_1" occurs the most whenever product_category_2 value is missing?

In [28]:
prod_2 = df[df['Product_Category_2'].isna()]
prod_1 = prod_2['Product_Category_1'].value_counts().index[0]
prod_1

5.0

#15.	Of all the users that exist in "test.csv", how many of them, also exist in "train.csv"

In [29]:
df2=pd.read_csv('/content/test.csv')

In [30]:
train_users = set(df['User_ID'].unique())
test_users = set(df2['User_ID'].unique())
common = train_users.intersection(test_users)
print("Number of users in both train and test datasets:",len(common)) 

Number of users in both train and test datasets: 5888


16.	Of all the users in "train.csv" how many of them also #exist in "test.csv"

In [31]:
train_users = set(df['User_ID'].unique())
test_users = set(df2['User_ID'].unique())
common = train_users.intersection(test_users)
print("Number of users in both train and test datasets:",len(common)) 

Number of users in both train and test datasets: 5888


#17.	What is the average purchase of customers who exist in "train.csv" but not in "test.csv"

In [36]:
unique_customer_ids_train = df[~df['User_ID'].isin(df2['User_ID'])]['User_ID'].unique()

In [37]:
rows = df[df['User_ID'].isin(unique_customer_ids_train)]


In [38]:
average = rows['Purchase'].mean()
average

nan

#18.	How many distinct combinations of "user_id" & "Product_id" are available in train.csv?

In [39]:
combinations = df[['User_ID','Product_ID']].nunique().sum()
combinations

9357

#20.	Write a function that takes variable name as input & gives out the frequency of occurrence table of the distinct values of the variable

In [40]:
def freq(var_name):
   
    freq = df[var_name].value_counts()
    
    freq = pd.DataFrame({'value': freq.index, 'frequency': freq.values}).sort_values('value')
    
    return freq

In [41]:
freq('Age')

value  frequency
6   0-17       5568
2  18-25      37308
0  26-35      81051
1  36-45      40491
3  46-50      16656
4  51-55      14259
5    55+       7949